core/modules/widgets/navigator.js

```diff
@@ -112,22 +107,43 @@ NavigatorWidget.prototype.replaceFirstTitleInStory = function(storyList,oldTitle
 		storyList.splice(0,0,newTitle);
 	}
 };
-
+	
 NavigatorWidget.prototype.addToStory = function(title,fromTitle) {
 	var storyList = this.getStoryList();
-	if(storyList) {
-		// See if the tiddler is already there
-		var slot = this.findTitleInStory(storyList,title,-1);
-		// If not we need to add it
-		if(slot === -1) {
-			// First we try to find the position of the story element we navigated from
-			slot = this.findTitleInStory(storyList,fromTitle,-1) + 1;
-			// Add the tiddler
-			storyList.splice(slot,0,title);
-			// Save the story
-			this.saveStoryList(storyList);
+	// Quit if we cannot get hold of the story list
+	if(!storyList) return;
+	// See if the tiddler is already there
+	var slot = storyList.indexOf(title);
+	// Quit if it already exists in the story river
+	if(slot >= 0) return;
+	// First we try to find the position of the story element we navigated from
+	var fromIndex = storyList.indexOf(fromTitle);
+	if(fromIndex >= 0) {
+		// How to open internal links that were clicked from *within* the story river?
+		var openLinkFromInsideRiver = $tw.wiki.getTiddlerText(
+			"$:/config/Navigation/openLinkFromInsideRiver", "below");
+		// The tiddler is added from inside the river
+		// Determine where to insert the tiddler; Fallback is "below"
+		switch(openLinkFromInsideRiver) {
+			case "top": slot = 0; break;
+			case "bottom": slot = storyList.length; break;
+			case "above": slot = fromIndex; break;
+			default: slot = fromIndex + 1;
 		}
+	} else {
+		// How to open links that were clicked from *outside* the story river?
+		var openLinkFromOutsideRiver = $tw.wiki.getTiddlerText(
+			"$:/config/Navigation/openLinkFromOutsideRiver", "top");
+		// The tiddler is opened from outside the river
+		// Determine where to insert the tiddler; Fallback is "top"
+			slot = (openLinkFromOutsideRiver === "bottom"
+							? storyList.length // Bottom
+							: 0); // Top
```

Comment:
Just to be clear, splitting this statement over multiple lines is reasonable, particularly so that you can interleave comments. I actually think it's frequently clearer to avoid the conditional operator and instead do something like this:

```
var slot = 0; // Default to top
if(openLinkFromOutsideRiver === "bottom") {
	slot = storyList.length; // Bottom
}
```

Generally, I value clarity more than brevity.